synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,659 @@
+            UserInfo: an object representing the user.
+        """
+        metadata = await self.load_metadata()
+        claims_params = {
+            "nonce": nonce,
+            "client_id": self._client_auth.client_id,
+        }
+        if "access_token" in token:
+            # If we got an `access_token`, there should be an `at_hash` claim
+            # in the `id_token` that we can check against.
+            claims_params["access_token"] = token["access_token"]
+            claims_cls = CodeIDToken
+        else:
+            claims_cls = ImplicitIDToken
+
+        alg_values = metadata.get("id_token_signing_alg_values_supported", ["RS256"])
+
+        jwt = JsonWebToken(alg_values)
+
+        claim_options = {"iss": {"values": [metadata["issuer"]]}}
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                token["id_token"],
+                key=jwk_set,
+                claims_cls=claims_cls,
+                claims_options=claim_options,
+                claims_params=claims_params,
+            )
+        except ValueError:
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                token["id_token"],
+                key=jwk_set,
+                claims_cls=claims_cls,
+                claims_options=claim_options,
+                claims_params=claims_params,
+            )
+
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+        return UserInfo(claims)
+
+    async def handle_redirect_request(
+        self, request, client_redirect_url: bytes
+    ) -> None:
+        """Handle an incoming request to /login/sso/redirect
+
+        It redirects the browser to the authorization endpoint with a few
+        parameters:
+
+          - ``client_id``: the client ID set in ``oidc_config.client_id``
+          - ``response_type``: ``code``
+          - ``redirect_uri``: the callback URL ; ``{base url}/_synapse/oidc/callback``
+          - ``scope``: the list of scopes set in ``oidc_config.scopes``
+          - ``state``: a random string
+          - ``nonce``: a random string
+
+        In addition to redirecting the client, we are setting a cookie with
+        a signed macaroon token containing the state, the nonce and the
+        client_redirect_url params. Those are then checked when the client
+        comes back from the provider.
+
+
+        Args:
+            request (SynapseRequest): the incoming request from the browser.
+                We'll respond to it with a redirect and a cookie.
+            client_redirect_url (bytes): the URL that we should redirect the
+                client to when everything is done
+        """
+
+        state = generate_token()
+        nonce = generate_token()
+
+        cookie = self._macaroon_generator.generate_oidc_session_token(
```

Comment:
Shouldn't this be `self._auth_handler`?

Reply:
Should it? I see a bunch of things related to sessions in the auth handler, but those sessions are persisted in-memory and not shared across workers, so a signed macaroon might be a more robust solution, as it is completely stateless server side. 

Or are you saying I should factor out the OIDC session generation/verification in the auth handler?

Reply:
Sorry my comment wasn't clear -- what I'm saying is that as far as I can tell this PR adds `generate_oidc_session_token` to the `AuthHandler` class, which gets stored as `self._auth_handler` unless I'm missing something.

Reply:
Doh, I misread the file and missed an additional class definition in the center. Please ignore this. 😄